A new test (dot1x) and a new tool (wpa) which was implemented to test 802.1X availability

krihal opened the pull request:
Initial pull request for 802.1X test and tool developed at SUNET Sweden. This was initially developed on the 4.3 branch using Python 2.

A few issues I can see right away:
-  The wpa tool is responsible of both doing authentication with WPA Supplicant and acquire an address using dhclient. We might want to split this into two different tools or rename it.

- This requires WPA Supplicant to be configured on the host system with D-Bus support enabled since we are using wpa_cli to configure things. I don't know how things like that is handled for other tools.

- When running this, you really want to run it in a separate network namespace, otherwise dhclient will replace the default route if an address is acquired and you risk loosing connectivity to the machine.

Any suggestions for improvements are very welcome.

/Kristofer - SUNET

mfeit-internet2 commented:
I've begun a review; expect a flurry of comments.

krihal commented:
Looking forward to it. :-)

mfeit-internet2 reviewed (changes requested):
This is headed in the right direction.  Pretty much everything I found in the code is minor.

Additionally:

Replace all occurrences of the old test name to the new one.  (There are a few stray references to `trace`.)

Our convention for naming JSON object members is to separate words with hyphens (e.g., `foo-bar` rather than `foo_bar`).

Name the test to be more-specific (e.g., `802.1x` or `ieee802.1x`).  Further comments on the test to follow in #1093.

If the `driver` can be determined based on system information about the `interface`, remove that parameter from the test and let the tool take care of it.

krihal commented:
> This is headed in the right direction. Pretty much everything I found in the code is minor.
> 
> Additionally:
> 
> Replace all occurrences of the old test name to the new one. (There are a few stray references to `trace`.)
> 
> Our convention for naming JSON object members is to separate words with hyphens (e.g., `foo-bar` rather than `foo_bar`).
> 
> Name the test to be more-specific (e.g., `802.1x` or `ieee802.1x`). Further comments on the test to follow in #1093.
> 
> If the `driver` can be determined based on system information about the `interface`, remove that parameter from the test and let the tool take care of it.

The test is now renamed to 802.1x. We don't use the driver, so I removed that parameter.

krihal commented on pscheduler-test-dot1x/dot1x/limit-is-valid:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python3
```
Updated to handle it the same way as netreach.

krihal commented on pscheduler-test-dot1x/dot1x/limit-passes:
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
```
Updated to handle it the same way as netreach.

krihal commented on pscheduler-test-dot1x/dot1x/result-format:
```diff
@@ -0,0 +1,63 @@
+
+
+if format == "text/plain":
+    print("*** Supplicant (Dot1X auth): %s" % json["supplicant"])
```
Fixed, output should now be formatted like your example.

krihal commented on pscheduler-test-dot1x/dot1x/validate.py:
```diff
@@ -0,0 +1,72 @@
+    return json_validate(json, schema)
+
+
+def limit_is_valid(json):
```
Fixed, function removed.

krihal commented on pscheduler-tool-wpa/wpa/Makefile:
```diff
@@ -0,0 +1,29 @@
+# Makefile for any tool class
+#
+
+NAME=wpa-supplicant
```
Renamed to wpa-supplicant.

krihal commented on pscheduler-tool-wpa/wpa/duration:
```diff
@@ -0,0 +1,19 @@
+import pscheduler
+
+json = pscheduler.json_load(exit_on_error=True)
+timeout_iso = json.get("timeout", "PT60S")
```
Adjusted it to more resonable 20 seconds.

krihal commented on pscheduler-tool-wpa/wpa/duration:
```diff
@@ -0,0 +1,19 @@
+
+json = pscheduler.json_load(exit_on_error=True)
+timeout_iso = json.get("timeout", "PT60S")
+timeout = pscheduler.iso8601_as_timedelta(timeout_iso) + datetime.timedelta(seconds=1)
```
Adjusted to 3 seconds, can't imagine we need more than that.

krihal commented on pscheduler-tool-wpa/wpa/run:
```diff
@@ -0,0 +1,236 @@
+    "succeeded": True,
+    "diags": "",
+    "error": None,
+    "result": {
```
Fixed.

krihal commented on pscheduler-tool-wpa/wpa/run:
```diff
@@ -0,0 +1,236 @@
+# supplicant is configured with DBUS support so we can use wpa_cli.
+#
+def wpa_add_network():
+    args = ["wpa_cli", "add_network"]
```
Now installing sudo configuration for both dhclient and wpa-supplicant.

krihal commented on pscheduler-tool-wpa/wpa/run:
```diff
@@ -0,0 +1,236 @@
+    logger.debug("Running wpa_cli: %s", " ".join(args))
+
+    try:
+        status, stdout, stderr = pscheduler.run_program(args)
```
Fixed.

krihal commented on pscheduler-tool-wpa/wpa/run:
```diff
@@ -0,0 +1,236 @@
+# Execute a command using wpa_cli and check if it was successful or
+# not.
+#
+def wpa_cmd(arg_str, response="OK"):
```
Fixed.

krihal commented on pscheduler-tool-wpa/wpa/run:
```diff
@@ -0,0 +1,236 @@
+#
+# Get status from WPA supplicant
+#
+def wpa_status():
```
Fixed.

krihal commented on pscheduler-tool-wpa/wpa/run:
```diff
@@ -0,0 +1,236 @@
+
+
+def dhcp_client(interface):
+    args_kill = ["sudo", "dhclient", "-x", interface]
```
Fixed.

krihal commented on pscheduler-tool-wpa/wpa/run:
```diff
@@ -0,0 +1,236 @@
+except KeyError:
+    pscheduler.fail("driver is a required parameter")
+
+phase2 = spec.get("phase2", "autheap=MSCHAPV2")
```
Totally missed a few things here so I cleaned that up and added two optional arguments, auth_inner and auth_outher to control this.

krihal commented on pscheduler-tool-wpa/wpa/run:
```diff
@@ -0,0 +1,236 @@
+except Exception as e:
+    logger.exception()
+    result["succeeded"] = False
+    result["result"]["dhclient"] = "Failed to acquire IP address"
```
Fixed.

krihal commented on pscheduler-tool-wpa/wpa/run:
```diff
@@ -0,0 +1,236 @@
+            result["result"]["error"] = "Acquired IP address even if dot1X authentication failed!"
+        else:
+            result["succeeded"] = True
+        result["result"]["dhclient"] = "Acquired IP address"
```
Same goes for supplicant, if we have bssid etc we can be sure that the authentication was successful.

krihal commented on pscheduler-test-dot1x/dot1x/validate.py:
```diff
@@ -0,0 +1,72 @@
+            "ip_addr":    {"$ref": "#/pScheduler/String"},
+        },
+        "required": [
+            # "duration",
```
Fixed.

krihal commented on pscheduler-test-dot1x/dot1x/validate.py:
```diff
@@ -0,0 +1,72 @@
+            "interface": {"$ref": "#/pScheduler/String"},
+            "driver":    {"$ref": "#/pScheduler/String"},
+            "eap_type":  {"$ref": "#/pScheduler/String"},
+            "username":  {"$ref": "#/pScheduler/String"},
```
Fixed.

krihal commented on pscheduler-test-dot1x/dot1x/validate.py:
```diff
@@ -0,0 +1,72 @@
+            "driver":    {"$ref": "#/pScheduler/String"},
+            "eap_type":  {"$ref": "#/pScheduler/String"},
+            "username":  {"$ref": "#/pScheduler/String"},
+            "password":  {"$ref": "#/pScheduler/String"},
```
Fixed.

krihal commented on pscheduler-test-dot1x/dot1x/validate.py:
```diff
@@ -0,0 +1,72 @@
+            "schema":    {"$ref": "#/pScheduler/Cardinal"},
+            "timeout":   {"$ref": "#/pScheduler/Duration"},
+            "interface": {"$ref": "#/pScheduler/String"},
+            "driver":    {"$ref": "#/pScheduler/String"},
```
Not sure what you mean here, can you explain a little further?

mfeit-internet2 replied:
Sure.  Anything that has a fixed set of valid values (e.g., `foo` and `bar` only) should be an enumerated string type that only allows those values.  See how the `rtt` test's `protocol` field is validated as an example.

krihal replied:
Thanks a lot, that helped. Pushed a fix for it.

epcjr commented:
You guys are probably way past this, but I thought I'd share anyway:

https://github.com/perfsonar/pscheduler/tree/master/scripts/PDK

There are some MD files there with instructions, as well as the templating
scripts.  Just FYI.

Thanks,

-Ed


On Fri, Dec 4, 2020 at 9:53 AM Mark Feit <notifications@github.com> wrote:

> *@mfeit-internet2* commented on this pull request.
> ------------------------------
>
> In pscheduler-test-dot1x/dot1x/validate.py
> <https://github.com/perfsonar/pscheduler/pull/1093#discussion_r536155581>:
>
> > +
> +#
> +# Validator for "trace" Test
> +#
> +
> +from pscheduler import json_validate
> +
> +
> +def spec_is_valid(json):
> +    schema = {
> +        "type": "object",
> +        "properties": {
> +            "schema":    {"$ref": "#/pScheduler/Cardinal"},
> +            "timeout":   {"$ref": "#/pScheduler/Duration"},
> +            "interface": {"$ref": "#/pScheduler/String"},
> +            "driver":    {"$ref": "#/pScheduler/String"},
>
> Sure. Anything that has a fixed set of valid values (e.g., foo and bar
> only) should be an enumerated string type that only allows those values.
> See how the rtt test's protocol field is validated as an example.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/perfsonar/pscheduler/pull/1093#discussion_r536155581>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLYWCYBNWZNOTMWXRPH533STDZXHANCNFSM4UKRGRSA>
> .
>

krihal commented:
I'll try to test this 

> You guys are probably way past this, but I thought I'd share anyway: https://github.com/perfsonar/pscheduler/tree/master/scripts/PDK There are some MD files there with instructions, as well as the templating scripts. Just FYI. Thanks,
> […](#)

Thanks for sharing, thats very useful. The plugin_dev script could have saved me quite some time.

krihal commented:
Thanks for the feedback, I've tried to fix it and improve thing in general. Also, I'll try to test this thoroughly to make sure everything is still working, there might be some regressions I've missed.

mfeit-internet2 commented:
Note for later:  4.4.0 will be making some minor changes to the file layout of packages that will have to be done for this package when it is pulled in.

mfeit-internet2 commented:
The UMich interns did work on #947, #1092 and #1167 that should cover most of this.

mfeit-internet2 commented:
Closing.   That was largely covered by the work UMich has done since.
